### DISCUSSION TITLE
How can I use `nixpkgs.stdenvAdapters.useMoldLinker` with `craneLib.devShell`?

### DISCUSSION THREAD

**Discussion by muni-corn** [ipetkov/crane#703] [Q&A] [answered]
I recently stumbled upon `crane` here and I am trying to wrap my head around how to create a development environment that can use `mold`. `nixpkgs` makes this easy with `stdenvAdapters`. This is my current setup:

```nix
devShells.default =
  pkgs.mkShell.override {
    stdenv = pkgs.stdenvAdapters.useMoldLinker pkgs.stdenv;
  } {
    packages =
      nativeBuildInputs
      ++ buildInputs
      ++ (with pkgs; [cargo-watch cargo-outdated]);
    LIBCLANG_PATH = "${pkgs.libclang.lib}/lib";
    LD_LIBRARY_PATH = pkgs.lib.makeLibraryPath buildInputs;
    # other env variables ...
  };
```

[(full flake here, currently using naersk)](https://codeberg.org/municorn/muni_bot/src/commit/7ee2bee387c0f1209ce55235ee6029654b6110cb/flake.nix)

but I can't simply replace `pkgs.mkShell` with `craneLib.devShell` because `stdenv` cannot be overridden in the same manner with crane.

**What's the best way to go about a similar setup with crane?** The goal is to have a development shell that compiles with rustc-codelift (done, easy with fenix) and uses `mold` for linking to speed up compile time with my Leptos project.

Lots of thanks in advance~!

**Comment by ipetkov** [comment, marked as answer]
Hi @muni-corn! Two insights which might be helpful here:

1. `craneLib.devShell` is basically a shortcut for including `rustc`/`cargo`/`clippy`/`rustfmt` from the toolchain declaration but it isn't strictly required; you can always use `pkgs.mkShell` directly (and pull in the toolchain override [if any] in there)
2. `craneLib.devShell` takes a `mkShell` input, so in theory you _could_ override it like so:

```nix
devShells.default = (craneLib.devShell.override {
  mkShell = pkgs.mkShell.override {
    stdenv = pkgs.stdenvAdapters.useMoldLinker pkgs.stdenv;
  };
}) {
  packages = [ /* ... */ ];
};
```

**Reply by muni-corn** [reply]
awesome! thank you so much! I'll give this a shot once I have free time again 😁

**Reply by ipetkov** [reply]
I've also added a note about this in the API docs so hopefully the next person who comes a long won't need to ask again :smile: 

https://github.com/ipetkov/crane/pull/709
